symengine/sets.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+RCP<const Set> Interval::set_union(const RCP<const Set> &o) const {
+    if (is_a<Interval>(*o)) {
+        const Interval &other = static_cast<const Interval &>(*o);
+        RCP<const Basic> start_start, end_end, m;
+        RCP<const Number> start, end;
+        bool left_open, right_open;
+        start_start = max({this->start_, other.start_});
+        end_end = min({this->end_, other.end_});
+        m = min({start_start, end_end});
+        if ((eq(*end_end, *start_start) and eq(*end_end, *m) and
+            ((eq(*end_end, *this->end_) and this->right_open_) or
+             (eq(*end_end, *other.end_) and other.right_open_))) or
+            (eq(*end_end, *m) and not eq (*end_end, *start_start))) {
+            throw std::runtime_error("not implemented");
+        } else {
+            if (eq(*min({this->start_, other.start_}), *this->start_))
+                start = this->start_;
+            else
+                start = other.start_;
+            if (eq(*max({this->end_, other.end_}), *this->end_))
+                end = this->end_;
+            else
+                end = other.end_;
+            left_open = ((neq(*this->start_, *start) or this->left_open_) and
+                             (neq(*other.start_, *start) or other.left_open_));
+            right_open = ((neq(*this->end_, *end) or this->right_open_) and
+                            (neq(*other.end_, *end) or other.right_open_));
+            return interval(start, end, left_open, right_open);
+        }
+    } else
+        return (*o).set_union(rcp_from_this_cast<const Set>());
+}
+
+bool Interval::is_subset(const RCP<const Set> &o) const {
+    if (is_a<Interval>(*o)) {
+        return this->__eq__(*this->set_intersection(o));
+    } else {
+        return (*o).is_superset(rcp_from_this_cast<const Set>());
+    }
+}
+
+bool Interval::is_proper_subset(const RCP<const Set> &o) const {
+    if (is_a<Interval>(*o)) {
+        return this->__eq__(*this->set_intersection(o)) and (not this->__eq__(*o));
+    } else {
+        return (*o).is_proper_superset(rcp_from_this_cast<const Set>());
+    }
+}
+
+bool Interval::is_superset(const RCP<const Set> &o) const {
+    return (*o).is_subset(rcp_from_this_cast<const Set>());
+}
+
+bool Interval::is_proper_superset(const RCP<const Set> &o) const {
+    return (*o).is_subset(rcp_from_this_cast<const Set>()) and (not __eq__(*o));
+}
+
+RCP<const Set> EmptySet::set_intersection(const RCP<const Set> &o) const {
+    return emptyset();
+}
+
+RCP<const Set> EmptySet::set_union(const RCP<const Set> &o) const {
+    return o;
+}
+
+std::size_t EmptySet::__hash__() const {
+    std::size_t seed = EMPTYSET;
+    return seed;
+}
+
+bool EmptySet::__eq__(const Basic &o) const {
+    if (is_a<EmptySet>(o))
+        return true;
+    return false;
+}
+
+int EmptySet::compare(const Basic &o) const {
+    if (is_a<EmptySet>(o))
+        return 0;
+    throw std::runtime_error("Not implemented");
```

Comment:
Checking that `o` is an `EmptySet` should be done only in Debug mode with `SYMENGINE_ASSERT`